### PR TITLE
fix(gateguard): deny dd whose input path is not word-initial

### PR DESCRIPTION
**Pull request by ntdat812**
## What Changed

`DESTRUCTIVE_SQL_DD` shared one trailing `\b` across every alternation arm:

```js
const DESTRUCTIVE_SQL_DD = /\b(drop\s+table|delete\s+from|truncate|dd\s+if=)\b/i;
```

`dd\s+if=` ends in `=`. A `\b` after a non-word character only holds when the **next** character is a word character, so the arm matched `dd if=x` and missed every path that starts with `/`, `.` or a quote. The word-boundary suffix now applies only to the arms that end in a word character:

```js
const DESTRUCTIVE_SQL_DD = /\b(?:drop\s+table|delete\s+from|truncate)\b|\bdd\s+if=/i;
```

This is the half of #2642 that @haelyra found still open on v2.1.0: the structural `findGitSubcommand()` parser already handles `git checkout -- .`, but the `dd` arm was untouched.

## Why This Change

Reproduced against current `main` (`d8409a4b`) — the classic disk-wipe spelling is allowed:

| command | before | after |
| --- | --- | --- |
| `dd if=/dev/zero of=/dev/sda` | **allowed** | denied |
| `dd if=./disk.img of=/dev/sdb` | **allowed** | denied |
| `dd if="/dev/zero" of=/dev/sda` | **allowed** | denied |
| `dd if=x` | denied | denied |
| `truncate table t` / `drop table users` / `delete from t` | denied | denied |
| `echo add if=1` | allowed | allowed |
| `echo truncated output` | allowed | allowed |

The change strictly widens the gate, and the split is what keeps the widening bounded: dropping the trailing `\b` outright would let `truncate` match `truncated`, and dropping the leading `\b` would let `dd if=` match inside `add if=`. Both spellings are in the tests.

## Testing Done

- [x] Manual testing completed
- [ ] Automated tests pass locally (`node tests/run-all.js`) — **see the note below; I could not complete this runner here and did not want to tick it**
- [x] Edge cases considered and tested

`tests/hooks/gateguard-fact-force.test.js` — 6 new cases, driven end-to-end through `run-with-flags.js` rather than against the regex, because the report is specifically that *the published hook* lets the slash-prefixed form through.

Reverting only `scripts/hooks/gateguard-fact-force.js` and keeping the tests:

```
✗ denies dd whose input path is not word-initial: dd if=/dev/zero of=/dev/sda
✗ denies dd whose input path is not word-initial: dd if=./disk.img of=/dev/sdb
✗ denies dd whose input path is not word-initial: dd if="/dev/zero" of=/dev/sda
✓ does not gate as destructive: echo add if=1
✓ does not gate as destructive: echo truncated output
✓ does not gate as destructive: git status
147 passed, 3 failed
```

With the fix: **150 passed, 0 failed**. The three "does not gate" cases pass either way by design — they guard against over-widening, they are not evidence of the bug.

Also checked by hand: `sudo dd if=/dev/zero` denied; `ddrescue if=…` and `xdd if=…` correctly not matched; `dd  if=` (double space) and `DD IF=` denied.

**Why `run-all.js` is unchecked:** it stops in my environment at `tests/lib/claude-plugin-setup.test.js`, which performs real marketplace `git clone`s — they fail here with `fatal: fetch-pack: invalid index-pack output`. That is environmental and unrelated to this change. `tests/hooks/gateguard-fact-force.test.js` is the only test file in the repo that references gateguard (`grep -rl gateguard tests/`), so the suite I did run covers this change's blast radius. Happy to post the full runner output if someone can point me at a way around the clone step.

## Type of Change

- [x] `fix:` Bug fix

## Security & Quality Checklist

- [x] No secrets or API keys committed
- [x] JSON files validate cleanly (no JSON touched)
- [x] Shell scripts pass shellcheck (no shell scripts touched)
- [x] No sensitive data exposed in logs or output
- [x] Follows conventional commits format

## One gap this does NOT close

`dd of=/dev/sda if=/dev/zero` — operands reversed — is still allowed, before and after this change, because the pattern requires `if=` immediately after `dd`. That is a separate defect from the `\b` reported here, and I did not want to widen a P0 gate's regex beyond the reported failure without a maintainer's call: something like `\bdd(?:\s+[a-z]+=\S*)*\s+if=` would cover it, but it changes the shape of the pattern rather than just its boundary handling, and a gate that blocks users' commands is the wrong place for me to guess. Say the word and I will add it here with its own cases, or file it separately.

---

*Written with Claude Code. The tables and test counts above are measured on `main` at `d8409a4b`, not estimated.*

### CODE REVIEW

**Comment by ntdat812** [comment]
@greptile-apps is right, and the finding is worth more than a narrow patch — thanks. Pushed 9249083e.

**Confirmed, with the caveat that the class was already there.** Measured against upstream `main`:

| command | main | first commit here | now |
| --- | --- | --- | --- |
| `echo dd if=x` | **denied** | denied | allowed |
| `echo dd if=/dev/zero` | allowed | **denied** | allowed |
| `grep dd if=/dev/zero file` | allowed | **denied** | allowed |

So the text match already gated `echo dd if=x` before I touched it; my boundary fix extended that to the slash and dot spellings. Either way the arm was deciding on where text appeared rather than on what runs, which is the actual defect — my PR body claimed "no new false positives" on the strength of `echo add if=1`, and that was the wrong case to test. Corrected.

**What changed.** `dd` moved out of `DESTRUCTIVE_SQL_DD` into `isDestructiveDd(tokens)`, beside `isDestructiveRm` and `isDestructiveGit`. The per-segment loop in `isDestructiveBash` already tokenizes every executable body, so the command word is available where the decision is made; the regex goes back to SQL only and is renamed `DESTRUCTIVE_SQL`.

Doing it structurally fixed three more things the text match could not:

| command | before | after |
| --- | --- | --- |
| `dd if=/dev/zero of=/dev/sda` | allowed | denied (the reported bug) |
| `sudo dd if=/dev/zero of=/dev/sda` | allowed | denied |
| `dd of=/dev/sda if=/dev/zero` | allowed | denied |
| `echo dd if=x` | denied | allowed |

The reversed-operand form is the gap I flagged at the bottom of the description as out of scope for a boundary fix. It is free once the check is token-based, so it is in.

Leading `sudo`/`doas`/`env`, their flags, and `VAR=value` prefixes are skipped so a wrapped invocation still resolves to `dd`. Flags are only skipped after a wrapper has been seen, so the scan cannot walk into an unrelated command's arguments.

**Tests:** 155 pass; **6 fail on upstream `main`** — the five deny cases plus `echo dd if=x`, which now pins the false positive so it cannot come back.

One thing I did not change: the file has no `sudo` handling anywhere, so `sudo rm -rf /` and `sudo git reset --hard` still bypass `isDestructiveRm`/`isDestructiveGit` on `main`. I only taught the new dd detector about wrappers rather than retrofitting the other two, since that is a wider behaviour change than this PR should carry. Happy to file it separately.

**Comment by ntdat812** [comment]
@coderabbitai both points are addressed — the critical one was real.

**Reversed and intervening operands (critical).** Fixed in 9249083e, the same rework that answered @greptile-apps: `dd` no longer matches as text at all. `isDestructiveDd(tokens)` resolves the command word first, then scans *every* remaining operand for `if=`, so operand order and intervening options stop mattering. Verified through the published hook:

| command | main | now |
| --- | --- | --- |
| `dd of=/dev/sda if=/dev/zero` | allowed | denied |
| `dd bs=1M if=/dev/zero of=/dev/sda` | allowed | denied |
| `sudo dd if=/dev/zero of=/dev/sda` | allowed | denied |

Both of your spellings are now in the regression list — `dd bs=1M if=…` added in 4ef558fd since you named it specifically.

**Vacuous allow-cases (minor).** Correct, and it was the weaker half of my own evidence: `if (output && output.hookSpecificOutput)` meant a crashed hook produced a green test. Fixed in f8d168be — the case asserts `result.code === 0` and that `parseOutput` returned something *before* reading the decision, with the `else` branch asserting pass-through echoes the input, matching the shape the existing retry case already uses.

156 pass; 6 fail on upstream `main`.

File: scripts/hooks/gateguard-fact-force.js [hunk 1]
```diff
@@ -375,6 +382,52 @@ function commandBasename(token) {
...
+    // an unrelated command's arguments.
+    if (sawWrapper && token.startsWith('-')) {
+      index += 1;
+      continue;
+    }
+    break;
+  }
+
+  if (index >= tokens.length || commandBasename(tokens[index]) !== 'dd') return false;
+  return tokens.slice(index + 1).some(operand => /^if=/i.test(operand));
```
[the remark this answers is not in the document]
Comment (ntdat812) [reply scripts/hooks/gateguard-fact-force.js:433]: Confirmed — the bypass is real on this branch. Measured against the built hook:

```
deny    dd if=/dev/zero of=/dev/sda
(none)  sudo -u root dd if=/dev/zero of=/dev/sda
deny    sudo dd if=/dev/zero of=/dev/sda
(none)  doas -u root dd if=/dev/zero of=/dev/sda
```

The fault is not in the dd detector — it is the wrapper walk skipping `-u` without consuming `root`, so `root` becomes the command word. That is exactly the subject of #2832, which replaces this ad-hoc loop with a shared `stripCommandWrappers()` carrying per-wrapper option arity (`sudo -p` takes a prompt, `command -p` does not).

I did not patch it here, because duplicating an arity table in `isDestructiveDd()` would then have to be deleted again. The two PRs also conflict textually in this exact function.

I built the combination locally to check it actually composes. With #2832 landed, `isDestructiveDd()` collapses to two lines — every call site already strips wrappers — and the reported form is gated:

```
deny    dd if=/dev/zero of=/dev/sda
deny    sudo -u root dd if=/dev/zero of=/dev/sda
deny    doas -u root dd if=/dev/zero of=/dev/sda
deny    env -u FOO dd if=/dev/zero of=/dev/sda
deny    env -S "dd if=/dev/zero of=/dev/sda"
deny    dd of=/dev/sda if=/dev/zero
(none)  echo dd if=/dev/zero
```

@affaan-m — happy to rebase this PR onto #2832 and drop the inline loop, or to fold the dd detector into #2832 and close this one, whichever you prefer. I did not push either without a decision on ordering.

File: scripts/hooks/gateguard-fact-force.js [hunk 2]
```diff
@@ -375,6 +382,52 @@ function commandBasename(token) {
...
+    if (/^[A-Za-z_][A-Za-z0-9_]*=/.test(token)) {
+      index += 1;
+      continue;
+    }
+    // Only skip flags once a wrapper has been seen, so this cannot walk past
+    // an unrelated command's arguments.
+    if (sawWrapper && token.startsWith('-')) {
+      index += 1;
+      continue;
+    }
```
[the remark this answers is not in the document]
Comment (ntdat812) [reply scripts/hooks/gateguard-fact-force.js:428]: Same bypass as the CodeRabbit thread above; confirmed by measurement and answered there. Short version: the option value is consumed correctly by the shared wrapper walk introduced in #2832, which supersedes the ad-hoc loop in this function — with both applied, `sudo -u root dd if=/dev/zero of=/dev/sda` denies, and so do the `doas -u`, `env -u`, and `env -S` spellings.

**Comment by ntdat812** [comment]
Rebased onto current `main`. One conflict, and it is worth naming because it looks like the fix landed upstream and it did not.

`main` renamed the regex to `DESTRUCTIVE_SQL_DD` and added `stripHeredocBodies()` ahead of it:

```js
const DESTRUCTIVE_SQL_DD = /\b(drop\s+table|delete\s+from|truncate|dd\s+if=)\b/i;
```

The `dd` arm still ends in `=` and still shares the trailing `\b`, so the boundary demands the *next* character be a word character. `dd if=/dev/zero` and `dd if=./img` still pass the guard; `dd if=x` is still denied. That is the reported defect, unchanged — only the constant's name moved.

Resolution keeps both halves rather than either side wholesale: `main`'s heredoc stripping stays (a phrase inside a heredoc body is data, not a command), and `dd` leaves the regex for the token-based `isDestructiveDd()`, which decides on `dd` being the command rather than on `dd if=` appearing anywhere in the line — so `echo dd if=/dev/zero` is no longer gated, `dd of=/dev/sda if=/dev/zero` is, and `sudo dd if=/dev/zero` still is.

Verification on the rebased branch:

```
node --test tests/hooks/gateguard-fact-force.test.js       1 pass, 0 fail

# with only scripts/hooks/gateguard-fact-force.js reverted to origin/main
node --test tests/hooks/gateguard-fact-force.test.js       0 pass, 1 fail
```

**Comment by chenhz01** [comment]
The `\b`-after-`=` analysis is exactly right, and moving `dd` to a token-based check is the correct fix — a text arm can't express "dd is the command". One destructive spelling still passes the new detector: `dd` with no `if=` operand reads stdin, so `dd of=/dev/sda bs=1M` or `cat /dev/zero | dd of=/dev/sda` destroys the target while never containing `if=`. Since the detector already resolves the command word and walks operands, extending it to flag any `of=` operand — or specifically `of=` pointing at an existing block device under `/dev/` — covers the stdin variant. The `sudo dd` wrapper resolution and order-free operand tests are good coverage; the same style would pin the new case (`dd of=/dev/sda`, `sudo dd bs=1M of=/dev/sda`). Want me to take a crack at it? Happy to open a PR.
